### ISSUE TITLE
2D immediate drawing and "z-order" by shape type

### ISSUE THREAD

**Issue by tintincastro** [james-j-obrien/bevy_vector_shapes#26]
Hello,

Bevy newbie here, I don't fully understand how `bevy_vector_shapes` works regarding to z-order in immediate mode. When using the painter, I expect that it will render my shapes in the order I draw them, but it doesn't seem to do so. Here's a basic example of what I mean :

```rust
use bevy::prelude::*;
use bevy_vector_shapes::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .insert_resource(ClearColor(Color::BLACK))
        .add_plugins(Shape2dPlugin::default())
        .add_systems(Startup, setup)
        .add_systems(Update, (draw, bevy::window::close_on_esc))
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn(Camera2dBundle::default());
}

fn draw(mut painter: ShapePainter) {
    painter.thickness = 8.;

    painter.hollow = true;
    painter.color = Color::YELLOW;
    painter.circle(150.);

    painter.translate(Vec3::X * 125.);

    painter.hollow = false;
    painter.color =  Color::BLACK;
    painter.rect(Vec2::ONE * 100.);

    painter.hollow = true;
    painter.color = Color::YELLOW;
    painter.rect(Vec2::ONE * 100.);

    painter.translate(Vec3::NEG_X * 125.);

    painter.hollow = true;
    painter.color = Color::WHITE;
    painter.circle(125.);
}
```

Based on the order of my painter commands, I would expect the previous code to output something like:
![expected](https://github.com/james-j-obrien/bevy_vector_shapes/assets/25725603/ffd796e5-d826-4c0d-8106-599c9755d182)

But here's what I get:
![actual](https://github.com/james-j-obrien/bevy_vector_shapes/assets/25725603/64d00b11-cfea-478c-b528-9847c13e4b2e)

No matter what I tried, I couldn't get the expected result (using reset(), using transforms, reverting the order I draw…). However, during my tests, I noticed that using the same shape type did not produce the same bug. For instance, this code:

```rust
// -- snip
fn draw(mut painter: ShapePainter) {
    painter.thickness = 8.;

    painter.hollow = true;
    painter.color = Color::YELLOW;
    painter.circle(150.);

    painter.translate(Vec3::X * 125.);

    painter.hollow = false;
    painter.color =  Color::BLACK;
    painter.circle(75.);

    painter.hollow = true;
    painter.color = Color::YELLOW;
    painter.circle(75.);

    painter.translate(Vec3::NEG_X * 125.);

    painter.hollow = true;
    painter.color = Color::WHITE;
    painter.circle(125.);
}
```
…outputs this:
![same-shape-type](https://github.com/james-j-obrien/bevy_vector_shapes/assets/25725603/d2dfe6cb-e872-4cac-b954-4fc279513f14)

…which led me to thinking that internally, bevy_shapes must collect shapes by type (rectangles, circles, lines…) for rendering, probably to render them in batches for performance reasons (I'm not savy enough yet to assert that, i'm just guessing at this stage…).

So my question is: is there any way to circumvent this behavior without impacting too much of the performance, in order to produce a more "natural" feeling with the way that shapes are drawn, or is this always gonna be this way?

I really hope not because the painter immediate mode is a real joy to use and I love how nicely the shapes renders to the screen. But since my project is just made of simple shapes drawns on top of each other, this quirk makes `bevy_vector_shapes` basically unusable in my case, and that makes me really really sad, because other than this issue, this crate would really be the perfect fit to me.

**Comment by james-j-obrien** [comment]
You are completely correct that this is caused by collecting the shapes together for batching. This is a known issue and I've worked to resolve it by contributing upstream to bevy. Those changes have been merged so in 0.12 this issue will be completely resolved and you will get the results you expect.

As a workaround in the meantime you can set the instance_id on ShapeConfig in order to break up the batches for a minor performance hit. If you set the circles on top to have a different id then those below the other shape you should get the results you expect. I don't like this solution as it's a bit cumbersome but it's the best I can do until 0.12 releases.

**Comment by james-j-obrien** [comment]
As a side note I really appreciate how well laid out this issue is, thanks for taking the time to do that. Hopefully 0.12 is ready for release soon :)

**Comment by tintincastro** [comment]
Wow, I'm so relieved to know that this issue will soon be a non-problem! Makes me looking forward for bevy 0.12 even more…

In the meantime, I will fiddle with `instance_id`, but don't bother to add options on my account, since it's already fixed for the next bevy release (to me, the appeal of `bevy_vector_shapes` lies in its simplicity, so I don't think there's a need to clutter the api for a temporary problem).

Thanks for replying so promptly. Have a nice day!

**Comment by james-j-obrien** [comment]
This is resolved in my bevy 0.12 branch: https://github.com/james-j-obrien/bevy_vector_shapes/tree/bevy-0.12.
